Deprecate ActiveSupport::Deprecation singleton usage

etiennebarrie opened the pull request:
### Motivation / Background

Following https://github.com/rails/rails/pull/46049, and discussions around Deprecation objects like https://github.com/rails/rails/pull/44772#issuecomment-1242511079:

> To me we should get rid of the singleton and just push the deprecation instance we today use in the singleton to the sub-framework's railties.

All the frameworks railties now have their own deprecator, but we were just missing the part where the singleton was deprecated.

cc @jonathanhefner 

### Detail

For each of the helper classes/users of `ActiveSupport::Deprecation#warn`, we add a deprecation warning about using them without passing a deprecator (giving a more specific deprecation instead of a generic one about using `ActiveSupport::Deprecation` directly).

Then the instance delegator deprecates its usage, giving a deprecation specific to the case: some things should be rewritten to use `Rails.application.deprecators` like `silence`, some directly to a `Deprecation` instance like `behavior`, and others like `warn` probably mean the user needs to create its own instance.

Finally, deprecates the singleton instance by de-documenting the inclusion of Singleton, and adding a deprecation warning when `ActiveSupport::Deprecation.instance` is used.

### Additional information

One thing I wasn't sure is if deprecated usage should still be documented or not even. I removed all the deprecated examples from the documentation.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

etiennebarrie committed:
Deprecate Module.deprecate without a deprecator

etiennebarrie committed:
Deprecate deprecate_constant without a deprecator

etiennebarrie committed:
Deprecate deprecation assertions without a deprecator

etiennebarrie committed:
Deprecate DeprecatedObjectProxy without a deprecator

etiennebarrie committed:
Deprecate DeprecatedInstanceVariableProxy without a deprecator

etiennebarrie committed:
Deprecate DeprecatedConstantProxy without a deprecator

etiennebarrie committed:
Document deprecators for library authors

zzak commented:
I _think_ we cannot remove documentation until **after** it is deprecated 🤔

etiennebarrie commented:
> I _think_ we cannot remove documentation until **after** it is deprecated 🤔

I checked and for example when `ActiveSupport::TimeWithZone#to_s(:db)` was deprecated in Rails 7.0, the documentation was removed as well. https://api.rubyonrails.org/v7.0/classes/ActiveSupport/TimeWithZone.html#method-i-to_s

Or the Enumerable.sum without a default value deprecation: afa83350deef906ed01b7acb5da6e9285aab661f

There are definitely cases where we mentioned the old deprecated way. https://api.rubyonrails.org/v6.1/classes/ActiveModel/Errors.html#method-i-each

It could go either way, either we really want to document the current behavior, even if it's deprecated, or we only want to document the non-deprecated behavior, and people can always go the previous version of the documentation to get the documentation.

Either way, if we come up with a decision on that I could go and document that in new deprecation cycle documentation: https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#breaking-changes

skipkayhil commented on activesupport/lib/active_support/core_ext/module/deprecation.rb:
```diff
@@ -1,25 +1,26 @@
   #
-  # \Custom deprecators must respond to <tt>deprecation_warning(deprecated_method_name, message, caller_backtrace)</tt>
-  # method where you can implement your custom warning behavior.
+  # Deprecators are typically instance of ActiveSupport::Deprecation, but you can also pass any object that responds to
```
```suggestion
  # Deprecators are typically an instance of ActiveSupport::Deprecation, but you can also pass any object that responds to
```

etiennebarrie replied:
I switched to use the singular instead.

skipkayhil commented on activesupport/lib/active_support/deprecation.rb:
```diff
@@ -3,8 +3,33 @@
+  # \Deprecation specifies the API used by Rails to deprecate methods, instance variables, objects, and constants. It's
+  # also available for gems or applications.
+  #
+  # For a gem, use Deprecation.new to create a Deprecation object, store it in your module or class (in order for users to be able
```
```suggestion
  # For a gem, use Deprecation.new to create a Deprecation object and store it in your module or class (in order for users to be able
```

skipkayhil commented on activesupport/lib/active_support/deprecation.rb:
```diff
@@ -3,8 +3,33 @@
+  #     end
+  #   end
+  #
+  # For a Railtie or Engine, you may also want to add it to the application's deprecators, in order for the application
+  # configuration to be applied to it.
```
```suggestion
  # For a Railtie or Engine, you may also want to add it to the application's deprecators, so that the application's
  # configuration can be applied to it.
```

skipkayhil commented on activesupport/lib/active_support/deprecation/proxy_wrappers.rb:
```diff
@@ -53,15 +53,15 @@ def warn(callstack, called, args)
+    # DeprecatedInstanceVariableProxy transforms an instance variable into a deprecated one. It takes an instance of a
+    # class, a method on that class, an instance variable, and a deprecator as the last argument.
+    #
+    # Trying to use the deprecated instance variable will result in a deprecation warning, pointing to the method as a
+    # replacement. This allows to change an API based on an instance variable access to a method call.
```
```suggestion
    # Trying to use the deprecated instance variable will result in a deprecation warning, pointing to the method as a
    # replacement.
```

I think the second sentence here is a little repetitive of the previous sentence.

skipkayhil commented on activesupport/lib/active_support/core_ext/module/deprecation.rb:
```diff
@@ -1,25 +1,26 @@
+  def deprecate(*method_names, deprecator: nil, **options)
+    if deprecator.is_a?(ActiveSupport::Deprecation)
+      deprecator.deprecate_methods(self, *method_names, **options)
+    elsif deprecator
+      ActiveSupport::Deprecation._instance.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
+    else
+      ActiveSupport.deprecator.warn("Module.deprecate without a deprecator is deprecated")
+    end
```
```suggestion
    else
      if deprecator.nil?
        ActiveSupport.deprecator.warn("Module.deprecate without a deprecator is deprecated")
        deprecator = ActiveSupport::Deprecation._instance
      end
      ActiveSupport::Deprecation._instance.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
    end
```

Shouldn't the last branch keep the old behavior (along with the warning)? Otherwise its effectively a removal instead of a deprecation

Maybe the `"deprecate_methods delegator is deprecated"` test should be updated to ensure that the deprecation is still in place as well?

skipkayhil reviewed (commented):
A few documentation suggestions and one implementation question, overall this looks good!

etiennebarrie committed:
Deprecate ActiveSupport::Deprecation instance delegation

All methods called on ActiveSupport::Deprecation are delegated to the
singleton instance, but now that each framework has its own Deprecation
instance, it's no longer necessary for internal deprecations.

For external code, instead of using ActiveSupport::Deprecation.warn,
they should create their own instance of the class and call warn on it.

etiennebarrie commented:
I've also noticed the documentation for collect_deprecations is incorrect and it was missing a test for `collect_deprecations` (#43774), so I've added a commit about that.

I realize this PR is too big, I'll create PRs for the things that can clearly be separated.

skipkayhil commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,26 @@
+    `config.active_support.report_deprecations` set to `false` in the production environment will also disable your
+    deprecator.
+
+    *Étienne Barrié*
```
```suggestion
    *Étienne Barrié*

```

etiennebarrie replied:
Oops bad rebase thanks!

skipkayhil reviewed (commented):
:bow: one more thing

matthewd commented on activesupport/lib/active_support/deprecation/instance_delegator.rb:
```diff
@@ -18,7 +17,27 @@ def include(included_module)
+          singleton_class.module_eval <<~RUBY, __FILE__, __LINE__ + 1
+            def #{method_name}(#{args})
+              ActiveSupport.deprecator.warn("Calling #{method_name} on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use #{use_instead} instead)")
+              ActiveSupport::Deprecation._instance.#{method_name}(#{args})
```
I think silence, behavior, etc should be redirected to `Rails.application.deprecators`, not `_instance`... otherwise they're going to immediately stop affecting all Rails-generated deprecations that have already been pointed elsewhere.

These lines also need to go the other way around, at minimum, to have some hope of directing the meta-deprecation warning towards the intended deprecation config.

I just discovered that my application had been failing to identify deprecation warnings for a while because `ActiveSupport::Deprecation.disallowed_warnings = :all` was no longer having the intended effect.

etiennebarrie replied:
Well, this PR only keeps the current main behavior and only adds the deprecation for calling `ActiveSupport::Deprecation.disallowed_warnings=`, but there is indeed a loss of behavior between `ActiveSupport::Deprecation.disallowed_warnings=` in 7.0 and here.

I do wonder how much we want to support this "undocumented behavior" though. For example https://api.rubyonrails.org/v7.0/classes/ActiveSupport/Deprecation/Disallowed.html mentions `ActiveSupport::Deprecation.disallowed_behavior`, but documents an instance writer for `disallowed_warnings`, which means it should have been used as `ActiveSupport::Deprecation.new.disallowed_warnings=`, or maybe `ActiveSupport::Deprecation.instance.disallowed_warnings=` (since `Singleton` was documented as an included module).

But the "correct" way was to set the configuration on `Rails.application.config.active_support`, and that has preserved the correct behavior in https://github.com/rails/rails/pull/46328.

There was a mention in the [Configuring guide](https://guides.rubyonrails.org/v7.0/configuring.html#config-active-support-disallowed-deprecation-warnings) to using the instance delegators as an alternative so we can definitely argue it was documented behavior.
> Alternatively, you can set `ActiveSupport::Deprecation.disallowed_warnings`

Anyway, I'll try to do the right thing and restore the 7.0 behavior of not only impacting the singleton instance but all deprecators when the instance delegators are used for these settings. The deprecation message correctly points how to get the previous behavior without a deprecation by setting on `Rails.application.deprecators`, so we should do the same.

jonathanhefner replied:
> I just discovered that my application had been failing to identify deprecation warnings for a while because `ActiveSupport::Deprecation.disallowed_warnings = :all` was no longer having the intended effect.

My apologies! :pray:  My (loose) reasoning was:

1. Previous deprecations that exist in 7.0 will (likely) be removed for 7.1.
2. New deprecations that have been added for 7.1 will require new `silence` blocks, `disallowed_warnings` values, etc for applications.
3. Since applications will already be updating `silence` blocks, `disallowed_warnings` values, etc to match the new 7.1 deprecations, it would be reasonable for them to switch from `ActiveSupport::Deprecation.*` to `Rails.application.deprecators.*` at the same time.

I did not take into account applications that are between 7.0 and 7.1 (i.e tracking `main` branch).

matthewd replied:
> Since applications will already be updating silence blocks, disallowed_warnings values, etc to match the new 7.1 deprecations

Even ignoring the running-on-main state, any upgrading-to-7.1 application is potentially currently configured in some way to direct/disallow deprecations. At absolute minimum, these meta-deprecations need to go into exactly that channel. 

The whole point of our deprecation system is that we _keep things functioning as they were before_, while warning that the application should be adapted to the new way. Creating a new parallel system next to where the old one was, and then reasoning that people will start using it because it's there, does not achieve that.

> New deprecations that have been added for 7.1 will require new silence blocks, disallowed_warnings values, etc for applications.

Specific disallows perhaps, but `:all` is pretty universal.

(I also think that we should still have a way to globally silence deprecations, or to globally disallow deprecations, without involving a railties dependency, but whatever.)

etiennebarrie committed:
Apply instance delegators to application deprecators when available

This restores the 7.0 behavior of the instance delegator impacting all
deprecations from Rails.

etiennebarrie committed:
Deprecate ActiveSupport::Deprecation.instance

etiennebarrie commented on activesupport/lib/active_support/core_ext/module/deprecation.rb:
```diff
@@ -1,25 +1,27 @@
+      ActiveSupport::Deprecation._instance.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
+    else
+      ActiveSupport.deprecator.warn("Module.deprecate without a deprecator is deprecated")
+      ActiveSupport::Deprecation.deprecate_methods(self, *method_names, **options)
```
This was the one I wanted to post "Great catch" on. I've changed the test to check there's only one deprecation and it showed that bug, fixed it.

jonathanhefner commented on activesupport/lib/active_support/deprecation/instance_delegator.rb:
```diff
@@ -18,7 +17,29 @@ def include(included_module)
+        end
+
+        def instance
+          ActiveSupport.deprecator.warn("ActiveSupport::Deprecation.instance is deprecated (use your own Deprecation object)")
+          super
```
Similarly, to cover e.g. `ActiveSupport::Deprecation.instance.behavior = ...`:

```suggestion
          super
          ensure
            ActiveSupport::Deprecation._instance.warn("ActiveSupport::Deprecation.instance is deprecated (use your own Deprecation object)")
```

Or, if we make `ActiveSupport.deprecator` == `ActiveSupport::Deprecation._instance`:

```suggestion
          super
          ensure
            ActiveSupport.deprecator.warn("ActiveSupport::Deprecation.instance is deprecated (use your own Deprecation object)")
```

jonathanhefner replied:
You're right — swapping the order here is nonsensical. :man_facepalming: :man_facepalming:

Using `ActiveSupport::Deprecation._instance` as the underlying deprecator will benefit subsequent calls though.  e.g.:

```ruby
ActiveSupport::Deprecation.instance.behavior = ...
ActiveSupport::Deprecation.instance.disallowed_behavior = ...
```

(Which is now covered since `ActiveSupport.deprecator` is `ActiveSupport::Deprecation._instance`.)

jonathanhefner commented on activesupport/lib/active_support/core_ext/module/deprecation.rb:
```diff
@@ -1,25 +1,27 @@
+    if deprecator.is_a?(ActiveSupport::Deprecation)
+      deprecator.deprecate_methods(self, *method_names, **options)
+    elsif deprecator
+      ActiveSupport::Deprecation._instance.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
```
We will probably remove `ActiveSupport::Deprecation._instance` whenever we remove `ActiveSupport::Deprecation.instance`.  So I think eventually this will need to be:

```suggestion
      ActiveSupport.deprecator.deprecate_methods(self, *method_names, **options, deprecator: deprecator)
```

jonathanhefner replied:
> I think there's another discussion to have about this duck-typed API (only implementing `deprecation_warning`), it's not very Rails-y, and does not offer much value since you can just pass a real object, I'm not sure why this both added a way to instantiate deprecators and the duck-typed API: [71993c6](https://github.com/rails/rails/commit/71993c6f9770b1350aa41fe8c68f1dd2c7800403)

I agree, though I might be missing something.  It seems like the intent was to make the API more "lightweight" so that the user doesn't have to be aware of the `ActiveSupport::Deprecation` class.  But, like you said, the value of that is questionable.  Also, if a lightweight variant is desirable, a callable (e.g. a `Proc`) seems like a better choice.

etiennebarrie committed:
Use the singleton instance as ActiveSupport deprecator

jonathanhefner commented on activesupport/lib/active_support/deprecation.rb:
```diff
@@ -3,8 +3,33 @@
 module ActiveSupport
-  # \Deprecation specifies the API used by Rails to deprecate methods, instance
-  # variables, objects, and constants.
+  # \Deprecation specifies the API used by Rails to deprecate methods, instance variables, objects, and constants. It's
+  # also available for gems or applications.
```
It's not a blocker, but keeping lines like this (not code examples) wrapped at 80-ish characters would be consistent with other method docs (for example, [further down](https://github.com/rails/rails/pull/47354/files#diff-2bfc6058d43af77dca7dbdd80a1e0cba368f7019ae493e615f9f5a70805738b7R60-R67)).

rafaelfranca commented on railties/lib/rails/deprecator.rb:
```diff
@@ -2,6 +2,6 @@
 module Rails
   def self.deprecator # :nodoc:
-    ActiveSupport::Deprecation.instance
+    @deprecator ||= ActiveSupport::Deprecation.new
```
I'm rereading https://github.com/rails/rails/pull/46049/files#r974695808, and now I just remember that maybe this one should be the instance otherwise we can't silence deprecation in the singleton anymore?

etiennebarrie replied:
We can silence deprecations in the singleton by using `ActiveSupport::Deprecation.silence` like it was possible in 7.0. That will cause a deprecation for the use of `silence` though.
In a Rails app, we can use `Rails.application.deprecators.silence` and it will silence the deprecation without also causing a deprecation, because `ActiveSupport.deprecator` (the singleton instance) is in the application deprecators as soon as the application is initialized.

jonathanhefner replied:
`Rails.deprecator` is the deprecator for the `railties` gem, similar to how `ActiveRecord.deprecator` is the deprecator for the `activerecord` gem.  It is currently used in a few places:

https://github.com/rails/rails/blob/a9506eb06243d309472071a5d65a3307d15580d4/railties/lib/rails/application/configuration.rb#L339-L347

https://github.com/rails/rails/blob/2d97136e38080edcb908d4b130b2fdda7c984b34/railties/lib/rails/generators/testing/behavior.rb#L112

https://github.com/rails/rails/blob/8542b748ba3eb257321fe0bc31742c72baf4975d/railties/lib/rails/commands/dbconsole/dbconsole_command.rb#L66-L72

However, we could use `ActiveSupport.deprecator` for those if you prefer.

With or without `Rails.deprecator`, `Rails.application.deprecators.silence` will still silence `ActiveSupport::Deprecation.warn`:

  ```irb
  irb> ActiveSupport::Deprecation.warn "boom"
  DEPRECATION WARNING: boom
  DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use your own Deprecation object instead)

  irb> Rails.application.deprecators.silence { ActiveSupport::Deprecation.warn "boom" }
  # No output
  ```

rafaelfranca replied:
If that is the railties deprecation we should rename it in the register to `deprecators[:railties]`, that was the point of my comment in the linked PR.

jonathanhefner replied:
@rafaelfranca Done in #47919.

ghiculescu commented on activesupport/lib/active_support/deprecation/instance_delegator.rb:
```diff
@@ -18,7 +17,30 @@ def include(included_module)
+              target = "(defined?(Rails.application.deprecators) ? Rails.application.deprecators : ActiveSupport::Deprecation._instance)"
+              "Rails.application.deprecators.#{method_name}"
+            when :warn, :deprecate_methods, :gem_name, :gem_name=, :deprecation_horizon, :deprecation_horizon=
+              "your own Deprecation object"
```
Could we give an example here? Consider this code in an app (not in a gem/engine), which I think is pretty common and there is lots of internet recommendations to use:

```ruby
ActiveSupport::Deprecation.warn "foo"
```

You get this warning:

```
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use your own Deprecation object instead)
```

I think a more useful warning would be:

```
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use `ActiveSupport.deprecator.warn` instead)
```

Unless we are recommending that apps define their own deprecators as well? (Which to me seems excessive for most apps.)

ghiculescu replied:
https://github.com/rails/rails/pull/47978

etiennebarrie replied:
I agree that we could give a better example of how to fix this deprecation. We absolutely do not want to point to `ActiveSupport.deprecator` though. The whole point of this deprecation cycle is to stop third-party code from generating deprecations that seem to come from Rails.

ghiculescu replied:
But what about first party code? What should it do?

etiennebarrie replied:
I meant third-party from the point of view of Rails, i.e. libraries and the application.

Regarding application deprecations, at Shopify, we've added an application deprecator in `config/application.rb`:

```ruby
module Shopify
  class Application < Rails::Application

    # ...

    initializer :deprecator do |app|
      app.deprecators[:shopify] = ActiveSupport::Deprecation.new("soon", "Shopify/shopify")
    end
  end
end
```

And then we use it directly:
```ruby
Rails.application.deprecators[:shopify].warn("some warning")
```

We've debated whether to do like a gem would typically do and have an API like `Shopify.deprecator` but ultimately felt like the out-of-the-box API from Rails was good enough.

I also thought that maybe Rails could always define an application deprecator to be used directly but didn't propose it because ultimately I don't think most apps need deprecations (it's possible to just change callers and callees at once, unlike in a library/framework), the deprecation horizon doesn't make much sense, etc.

If an application needs a deprecator, it can do what we did above, or do like in a gem, define a singleton method on the application module, set in the application deprecators in an initializer.

I'm not sure what we should write in the deprecation to make that clearer, there's some setup involved… Yes the deprecated API was clean and nice, but in the end it was misguided as separate gems and applications in addition to Rails were all sharing the same `Deprecation` instance.

ghiculescu replied:
Hmm. I don't think the fact that setup is required should be celebrated. I understand Shopify doing this but for the long tail of small apps I think this is overly complicated vs. what Rails 7 and before offered.
